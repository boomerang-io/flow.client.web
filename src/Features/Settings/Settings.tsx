import React from "react";
import { Helmet } from "react-helmet";
import { useQuery, useMutation, useQueryClient } from "react-query";
import { Box } from "reflexbox";
import {
  Accordion,
  ErrorMessage,
  FeatureHeader as Header,
  FeatureHeaderTitle as HeaderTitle,
  FeatureHeaderSubtitle as HeaderSubtitle,
  SkeletonPlaceholder,
  notify,
  ToastNotification,
} from "@carbon/react";
import SettingsSection from "./SettingsSection";
import sortBy from "lodash/sortBy";
import EmptyState from "Components/EmptyState";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { DataDrivenInput } from "Types";
import styles from "./settings.module.scss";

export type SettingsGroup = {
  description: string;
  name: string;
  key: string;
  config: DataDrivenInput[];
};

const platformSettingsUrl = serviceUrl.resourceSettings();

const FeatureLayout: React.FC = ({ children }) => {
  return (
    <>
      <Header
        className={styles.header}
        includeBorder={false}
        header={
          <>
            <HeaderTitle className={styles.headerTitle}>Settings</HeaderTitle>
            <HeaderSubtitle>Adjust Flow settings</HeaderSubtitle>
          </>
        }
      />
      <Box p="2rem" overflowY="auto" className={styles.container}>
        {children}
      </Box>
    </>
  );
};

const Settings: React.FC = () => {
  const queryClient = useQueryClient();

  const { data, error, isLoading } = useQuery({
    queryKey: platformSettingsUrl,
    queryFn: resolver.query(platformSettingsUrl),
  });

  const { mutateAsync: updateSettingMutator } = useMutation(resolver.putPlatformSettings, {
    onSuccess: () => queryClient.invalidateQueries(platformSettingsUrl),
  });

  const handleOnSave = async (
    values: { [key: string]: any },
    settingsGroup: SettingsGroup,
    setFieldError: (key: string, value: string) => void
  ) => {
    const newConfig = settingsGroup.config.map((input: any) => ({ ...input, value: values[input.key] }));
    const requestBody = [{ ...settingsGroup, config: newConfig }];
    try {
      await updateSettingMutator({ body: requestBody });
      notify(<ToastNotification title="Update Settings" subtitle="Settings succesfully updated" kind="success" />);
      setFieldError("initialerror", "required");
    } catch (e) {
      notify(<ToastNotification title="Something's Wrong" subtitle="Request to update settings failed" kind="error" />);
    }
  };

  if (isLoading) {
    return (
      <FeatureLayout>
        <Box className={styles.skeletonPlacholderContainer}>
          <SkeletonPlaceholder />
          <SkeletonPlaceholder />
          <SkeletonPlaceholder />
          <SkeletonPlaceholder />
          <SkeletonPlaceholder />
        </Box>
      </FeatureLayout>
    );
  }

  if (error) {
    return (
      <FeatureLayout>
        <ErrorMessage />
      </FeatureLayout>
    );
  }

  const sortedPlatformSettings = sortBy(data, (settingObj) => settingObj.name);
  return (
    <FeatureLayout>
      <Helmet>
        <title>Settings</title>
      </Helmet>
      {!sortedPlatformSettings.length ? (
        <EmptyState />
      ) : (
        <Accordion>
          {sortedPlatformSettings.map((settingsGroup, index) => (
            <SettingsSection index={index} key={index} onSave={handleOnSave} settingsGroup={settingsGroup} />
          ))}
        </Accordion>
      )}
    </FeatureLayout>
  );
};

export default Settings;
