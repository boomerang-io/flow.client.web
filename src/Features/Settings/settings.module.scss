.header {
  h1 {
    margin: 0;
  }
}

.container {
  width: 100%;
  overflow-y: auto;

  :global .cds--accordion__content {
    padding-right: 1rem;
    padding-top: 3rem;
  }
  :global .cds--bmrg-dynamic-formik {
    align-items: flex-start;
    display: flex;
    flex-flow: row wrap;
    justify-content: flex-start;
    position: relative;
  }
  :global .cds--bmrg-text-input {
    display: flex;
    flex: 0 0 50%;
    position: relative;
    min-width: 30rem;
    padding: 1rem 1rem 0;
  }
  :global .cds--dropdown__wrapper,
  :global .cds--bmrg-select,
  :global .cds--bmrg-multi-select {
    display: flex;
    flex: 0 0 50%;
    flex-direction: column;
    left: 1rem;
    position: relative;
    min-width: 30rem;
    padding: 1rem 2rem 0 0;
  }
  :global .cds--bmrg-toggle__container {
    flex: 0 0 50%;
    max-width: 50%;
    padding: 1rem 1rem 0;
  }

  :global .cds--accordion__heading {
    align-items: center;
  }
}

.skeletonPlacholderContainer {
  * {
    width: 100%;
  }

  > * + * {
    margin-top: 1rem;
  }

  :first-child {
    height: 14rem;
  }

  :not(:first-child) {
    height: 3rem;
    min-height: 3rem;
  }
}
