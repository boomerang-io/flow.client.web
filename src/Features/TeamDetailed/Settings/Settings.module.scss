@import "Styles/mixins";

.settingsContainer {
  padding: 1.5rem 2rem;
  position: relative;
  overflow: auto;

  :global .cds--bmrg-toggle {
    margin-bottom: 2rem;
  }

  :global .cds--label {
    color: var(--cds-text-01);
    font-size: 0.875rem;
    margin-bottom: 1rem !important;
  }

  :global .cds--form__helper-text {
    align-items: center;
    display: flex;
    color: var(--cds-text-01);
    font-size: 0.75rem;
    font-style: normal !important;
    line-height: 0.9375rem;
    margin-bottom: 1rem !important;
    max-width: 100%;
  }
}

.editTeamNameContainer {
  margin: 2rem 0;
}

.actionableNameContainer {
  display: flex;
  align-items: center;
}

.teamNameLabel {
  color: var(--cds-icon-02);
  font-size: 0.75rem;
  margin-bottom: 0.5rem;
}

.teamEditIcon {
  cursor: pointer;
  height: 1rem;
  margin-left: 1rem;
  &:focus {
    @include focus-outline();
  }
  path,
  rect {
    fill: var(--cds-interactive-01);
  }
}

.teamNameModalContainer {
  max-height: 22rem;

  :global .cds--modal-content {
    padding-right: 1rem;
  }
}

.detailedListContainer {
  margin-top: 0.5rem;
  width: 60rem;
}

.detailedListParagraph {
  color: var(--cds-text-02);
  font-size: 0.75rem;
  font-style: normal !important;
  line-height: 0.9375rem;
  margin-bottom: 1rem !important;
  max-width: 41.2rem;
}

.detailedListGrid {
  display: grid;
  column-gap: 1rem;
  grid-template-columns: repeat(auto-fill, 20.6rem);
  width: 43rem;
}

.detailedListGridItem {
  width: 20.6rem;
}

.detailedListTitle {
  color: var(--cds-text-02);
  font-size: 0.75rem;
  letter-spacing: 0.02rem;
  line-height: 0.9375rem;
  margin-top: 1.5rem;
  margin-bottom: 0.5rem;
  @include text-overflow-ellipsis();
}

.detailedListDescription {
  font-size: 0.875rem;
  letter-spacing: 0.01rem;
  line-height: 1.125rem;
  max-width: 41.2rem;
  @include multiline-text-overflow-ellipsis();
}

.detailedListDescriptionEmpty {
  color: var(--cds-text-02);
  font-size: 0.875rem;
  font-style: italic;
  letter-spacing: 0.01rem;
  line-height: 1.125rem;
  max-width: 41.2rem;
}

.editButton {
  justify-content: center;
  width: 2rem;

  :global .bx--btn__icon {
    margin: 0;
    height: 0.875rem;
  }
}

.notificationsContainer {
  align-items: center;
  display: flex;
  justify-content: space-between;
  margin-bottom: 1rem;
}

.settingsDescription {
  font-size: 0.875rem;
  line-height: 1.125rem;
  max-width: 41.25rem;
}

.sectionContainer {
  border-top: 0.0625rem solid var(--cds-active-ui);
  margin-top: 1.5rem;
  padding-top: 1.5rem;
  width: 100%;
}

.sectionHeader {
  align-items: center;
  display: flex;
}

.sectionTitle {
  font-size: 1.125rem;
  letter-spacing: 0;
  line-height: 1.4375rem;
  margin: 0 0.5rem 0 0;
}

.sectionDescription {
  font-size: 0.875rem;
  letter-spacing: 0.01rem;
  line-height: 1.125rem;
  margin-top: 1.25rem;
  width: 41.1875rem;
}

.buttonWithMessageContainer {
  margin-top: 1.5rem;
  width: 41.1875rem;
}

.buttonHelperText {
  color: var(--cds-text-02);
  font-size: 0.75rem;
  font-style: normal !important;
  line-height: 0.9375rem;
  margin-bottom: 1rem !important;
  max-width: 41.2rem;
}

.closeTeamButton {
  margin-left: -0.5rem;
  padding: 0.1875rem 0.5rem !important;
}

.structuredListWrapper {
  max-width: 75rem;
  margin-top:0.25rem;
}

.structuredListHeader {
  display: flex;
}

.structuredListHeaderHoverIcon {
  margin-left: 0.25rem;
}