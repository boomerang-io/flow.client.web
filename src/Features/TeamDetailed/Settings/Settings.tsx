import React from "react";
import { Helmet } from "react-helmet";
import { useMutation, useQueryClient } from "react-query";
import { InlineNotification, Button } from "@carbon/react";
import {
  ConfirmModal,
  ComposedModal,
  notify,
  ToastNotification,
  TooltipHover,
} from "@boomerang-io/carbon-addons-boomerang-react";
import UpdateTeamName from "./UpdateTeamName";
import {
  StructuredListWrapper,
  StructuredListHead,
  StructuredListRow,
  StructuredListCell,
  StructuredListBody,
} from "@carbon/react";
import { Edit, Close, TrashCan, Add } from "@carbon/react/icons";
import sortBy from "lodash/sortBy";
import { FlowTeam } from "Types";
import LabelModal from "Components/LabelModal";
import styles from "./Settings.module.scss";
import { resolver, serviceUrl } from "Config/servicesConfig";

interface Label {
  key: string;
  value: string;
}

export default function Settings({ team, canEdit }: { team: FlowTeam; canEdit: boolean }) {
  const queryClient = useQueryClient();

  const patchTeamMutator = useMutation(resolver.patchUpdateTeam);

  const handleRemoveTeam = async () => {
    try {
      await patchTeamMutator.mutateAsync({ team: team.name, body: { isActive: false } });
      queryClient.invalidateQueries(serviceUrl.resourceTeam({ team: team.name }));
      notify(
        <ToastNotification title="Remove Team" subtitle={`Request to close ${team.name} successful`} kind="success" />
      );
    } catch (error) {
      // noop
    }
  };

  const handleAddLabel = async (value: Label) => {
    console.log("handleAddLabel", value);
    const newLabels = [...teamLabels, value];
    const newLabelsRecord = newLabels.reduce((acc, label) => {
      acc[label.key] = label.value;
      return acc;
    }, {} as Record<string, string>);

    try {
      await patchTeamMutator.mutateAsync({ team: team.name, body: { labels: newLabelsRecord } });
      queryClient.invalidateQueries(serviceUrl.resourceTeam({ team: team.name }));
      notify(
        <ToastNotification title="Remove Team" subtitle={`Request to close ${team.name} successful`} kind="success" />
      );
    } catch (error) {
      // noop
    }
  };

  const handleRemoveLabel = async (value: Label) => {
    console.log("handleRemoveLabel", value);

    const newLabels = teamLabels.filter((label) => label.key !== value.key);
    const newLabelsRecord = newLabels.reduce((acc, label) => {
      acc[label.key] = label.value;
      return acc;
    }, {} as Record<string, string>);

    console.log("newLabels", newLabelsRecord);

    try {
      await patchTeamMutator.mutateAsync({ team: team.name, body: { labels: newLabelsRecord } });
      queryClient.invalidateQueries(serviceUrl.resourceTeam({ team: team.name }));
      notify(
        <ToastNotification title="Remove Team" subtitle={`Request to close ${team.name} successful`} kind="success" />
      );
    } catch (error) {
      // noop
    }
  };

  // Convert Record/Map of Labels to Array of Label Object
  const teamLabels = team.labels ? Object.entries(team.labels).map(([key, value]) => ({ key, value })) : [];
  const labelsKeys = team.labels ? Object.keys(team.labels) : [];

  return (
    <section aria-label="Team Settings" className={styles.settingsContainer}>
      <Helmet>
        <title>{`Settings - ${team.name}`}</title>
      </Helmet>
      {!canEdit ? (
        <section className={styles.notificationsContainer}>
          <InlineNotification
            lowContrast
            hideCloseButton={true}
            kind="info"
            title="Read-only"
            subtitle="You don’t have permission to change these settings, but you can still see what’s going on behind the
            scenes."
          />
        </section>
      ) : null}
      <p className={styles.settingsDescription}>
        Configurable settings for this Team – Team Owners & Admins have the ability to make changes here.
      </p>
      <SettingSection
        title="Basic details"
        editModal={
          <ComposedModal
            composedModalProps={{
              containerClassName: styles.teamNameModalContainer,
            }}
            modalHeaderProps={{
              title: "Change team name",
              //   subtitle:
              //     "Try to keep it concise to avoid truncation in the sidebar. You must make sure the name is valid before it can be updated.",
            }}
            modalTrigger={({ openModal }: { openModal: () => void }) => (
              <button
                disabled={!canEdit}
                className={styles.teamEditIcon}
                onClick={openModal}
                data-testid="open-change-name-modal"
              >
                <Edit />
              </button>
            )}
          >
            {({ closeModal }: { closeModal: () => void }) => <UpdateTeamName closeModal={closeModal} team={team} />}
          </ComposedModal>
        }
      >
        <dl className={styles.detailedListContainer}>
          <div className={styles.detailedListGrid}>
            <div className={styles.detailedListGridItem}>
              <dt className={styles.detailedListTitle}>Display Name</dt>
              <dd className={styles.detailedListDescription}>{team.displayName}</dd>
            </div>
            <div className={styles.detailedListGridItem}>
              <dt className={styles.detailedListTitle}>Unique Identifier Name</dt>
              <dd className={styles.detailedListDescription}>{team.name}</dd>
            </div>
          </div>
        </dl>
      </SettingSection>
      <SettingSection title="Labels">
        <dl className={styles.detailedListContainer}>
          <p className={styles.detailedListParagraph}>Create custom labels can be useful when querying the API.</p>
          <StructuredListWrapper
            className={styles.structuredListWrapper}
            ariaLabel="Structured list"
            isCondensed={true}
          >
            <StructuredListHead>
              <StructuredListRow head>
                <StructuredListCell head>Key</StructuredListCell>
                <StructuredListCell head>Value</StructuredListCell>
                <StructuredListCell head />
              </StructuredListRow>
            </StructuredListHead>
            <StructuredListBody>
              {sortBy(teamLabels, "key").map((label: Label) => {
                const labelIndex = teamLabels.findIndex((labelFromList) => labelFromList.key === label.key);
                return (
                  <StructuredListRow key={label.key}>
                    <StructuredListCell className={styles.labelKeyCell}>{label.key}</StructuredListCell>
                    <StructuredListCell>{label.value}</StructuredListCell>
                    {canEdit && (
                      <>
                        <StructuredListCell>
                          <LabelModal
                            action={handleAddLabel}
                            isEdit
                            labelsKeys={labelsKeys.filter((labelKey) => labelKey !== label.key)}
                            selectedLabel={label}
                            modalTrigger={({ openModal }: { openModal: Function }) => (
                              <Button
                                kind="ghost"
                                iconDescription="edit label"
                                renderIcon={Edit}
                                size="sm"
                                onClick={openModal}
                              >
                                Edit
                              </Button>
                            )}
                          />
                          <ConfirmModal
                            modalTrigger={({ openModal }) => (
                              <Button
                                kind="danger--ghost"
                                iconDescription="delete label"
                                renderIcon={TrashCan}
                                size="sm"
                                onClick={openModal}
                                data-testid={`delete-token-button-${label}`}
                              />
                            )}
                            affirmativeAction={() => handleRemoveLabel(label)}
                            affirmativeButtonProps={{ kind: "danger" }}
                            affirmativeText="Yes"
                            negativeText="No"
                            title={`Are you sure?`}
                          >
                            Delete
                          </ConfirmModal>
                        </StructuredListCell>
                      </>
                    )}
                  </StructuredListRow>
                );
              })}
              <LabelModal
                action={handleAddLabel}
                labelsKeys={labelsKeys}
                modalTrigger={({ openModal }: { openModal: Function }) => (
                  <Button kind="ghost" iconDescription="add a new label" renderIcon={Add} size="md" onClick={openModal}>
                    Add a new label
                  </Button>
                )}
              />
            </StructuredListBody>
          </StructuredListWrapper>
        </dl>
        {/* <CreateToken getTokensUrl={getTokensUrl} principal={user.id} type="user" /> */}
      </SettingSection>
      <SettingSection title="Close Team">
        <div className={styles.buttonWithMessageContainer}>
          <p className={styles.buttonHelperText}>
            Done with your work here? Closing the team means its members will no longer be able to access this team, and
            access to the Services will be revoked.
          </p>
          <ConfirmModal
            affirmativeAction={() => handleRemoveTeam()}
            affirmativeButtonProps={{ kind: "danger", "data-testid": "confirm-close-team" }}
            title={`Close ${team.name}?`}
            negativeText="Cancel"
            affirmativeText="Close"
            modalTrigger={({ openModal }: { openModal: () => void }) => (
              <Button
                disabled={!canEdit}
                iconDescription="Close"
                kind="danger--ghost"
                onClick={openModal}
                renderIcon={Close}
                size="md"
                data-testid="close-team"
              >
                Close Team
              </Button>
            )}
          >
            Closing a team will submit a "leave team" request for each user on the team. After the requests are
            processed, the team will become "inactive". Are you sure you want to do this?
          </ConfirmModal>
        </div>
      </SettingSection>
    </section>
  );
}

function SettingSection({ children, description, editModal, title }) {
  return (
    <section className={styles.sectionContainer}>
      <div className={styles.sectionHeader}>
        <h1 className={styles.sectionTitle}>{title}</h1>
        {editModal}
      </div>
      {description ? <p className={styles.sectionDescription}>{description}</p> : null}
      {children}
    </section>
  );
}
