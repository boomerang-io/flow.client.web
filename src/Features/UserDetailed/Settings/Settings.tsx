import React from "react";
import { Helmet } from "react-helmet";
import { Edit } from "@carbon/react/icons";
import { Tag, Button } from "@carbon/react";
import { FlowUser } from "Types";
import styles from "./Settings.module.scss";

interface UserSettingsProps {
  user: FlowUser;
  userManagementEnabled: any;
}

export default function Settings({ user, userManagementEnabled }: UserSettingsProps) {
  const canEdit = userManagementEnabled;

  return (
    <section aria-label="User Settings" className={styles.settingsContainer}>
      <Helmet>
        <title>{`Settings - ${user?.name}`}</title>
      </Helmet>
      {!canEdit ? (
        <section className={styles.readOnly}>
          <Tag className={styles.readOnlyTag}>Read-only</Tag>
          <p className={styles.readOnlyText}>
            You don’t have permission to change any of these settings, but you can still see what’s going on behind the
            scenes.
          </p>
        </section>
      ) : (
        <p className={styles.settingsDescription}>
          These are configurable settings for this User – only Admins have the ability to make changes here
        </p>
      )}
      <SettingSection title="Basic details">
        <dl className={styles.detailedListContainer}>
          <div className={styles.detailedListGrid}>
            <div className={styles.detailedListGridItem}>
              <dt className={styles.detailedListTitle}>Name</dt>
              <dd className={styles.detailedListDescription}>{user.name}</dd>
            </div>
            <div className={styles.detailedListGridItem}>
              <dt className={styles.detailedListTitle}>Email</dt>
              <dd className={styles.detailedListDescription}>{user.email}</dd>
            </div>
          </div>
          {/* <dt className={styles.detailedListTitle}>Team Description</dt>
          {team.description ? (
            <dd className={styles.detailedListDescription}>{team.description}</dd>
          ) : (
            <dd className={styles.detailedListDescriptionEmpty}>
              No description. Add a brief description about this Team, which will appear in the header of this team
              page.
            </dd>
          )}
          <dt className={styles.detailedListTitle}>Logo</dt>
          {team.logo?.id && showLogo ? (
            <>
              <dd className={styles.detailedListDescription} style={{ marginBottom: "1rem" }}>
                {team.logo?.name ?? "---"}
              </dd>
              {teamLogo}
            </>
          ) : (
            <dd className={styles.detailedListDescriptionEmpty}>
              No logo. Upload a small image that will appear in the header of this team page. Useful if this team is
              specific to a client or sector, to help team members recognize the Team alignment at a glance.
            </dd>
          )} */}
        </dl>
      </SettingSection>
    </section>
  );
}

// function EditButton({ openModal }: any) {
//   return (
//     <Button
//       className={styles.editButton}
//       data-testid="settings-edit-button"
//       iconDescription="edit"
//       kind="ghost"
//       size="small"
//       renderIcon={Edit}
//       onClick={openModal}
//     />
//   );
// }

function SettingSection({ children, description, editModal, title }: any) {
  return (
    <section className={styles.sectionContainer}>
      <div className={styles.sectionHeader}>
        <h1 className={styles.sectionTitle}>{title}</h1>
        {editModal}
      </div>
      {description ? <p className={styles.sectionDescription}>{description}</p> : null}
      {children}
    </section>
  );
}
