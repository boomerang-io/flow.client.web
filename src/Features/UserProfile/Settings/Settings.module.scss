@import "Styles/mixins";

.settingsContainer {
  padding: 1.5rem 2rem;
  position: relative;

  :global .cds--bmrg-toggle {
    margin-bottom: 2rem;
  }

  :global .cds--label {
    color: var(--cds-text-01);
    font-size: 0.875rem;
    margin-bottom: 1rem !important;
  }

  :global .cds--form__helper-text {
    align-items: center;
    display: flex;
    color: var(--cds-text-01);
    font-size: 0.75rem;
    font-style: normal !important;
    line-height: 0.9375rem;
    margin-bottom: 1rem !important;
    max-width: 100%;
  }
}

.editTeamNameContainer {
  margin: 2rem 0;
}

.actionableNameContainer {
  display: flex;
  align-items: center;
}

.teamNameLabel {
  color: var(--cds-icon-02);
  font-size: 0.75rem;
  margin-bottom: 0.5rem;
}

.teamEditIcon {
  cursor: pointer;
  height: 1rem;
  margin-left: 1rem;
  &:focus {
    @include focus-outline();
  }
  path,
  rect {
    fill: var(--cds-interactive-01);
  }
}

.teamNameModalContainer {
  max-height: 22rem;

  :global .cds--modal-content {
    padding-right: 1rem;
  }
}

.detailedListGrid {
  display: grid;
  column-gap: 1rem;
  grid-template-columns: repeat(auto-fill, 20.6rem);
  width: 43rem;
}

.detailedListGridItem {
  width: 20.6rem;
}

.detailedListTitle {
  color: var(--cds-text-02);
  font-size: 0.75rem;
  letter-spacing: 0.02rem;
  line-height: 0.9375rem;
  margin-top: 1.5rem;
  margin-bottom: 0.5rem;
  @include text-overflow-ellipsis();
}

.detailedListDescription {
  font-size: 0.875rem;
  letter-spacing: 0.01rem;
  line-height: 1.125rem;
  max-width: 41.2rem;
  @include multiline-text-overflow-ellipsis();
}

.detailedListDescriptionEmpty {
  color: var(--cds-text-02);
  font-size: 0.875rem;
  font-style: italic;
  letter-spacing: 0.01rem;
  line-height: 1.125rem;
  max-width: 41.2rem;
}

.editButton {
  justify-content: center;
  width: 2rem;

  :global .bx--btn__icon {
    margin: 0;
    height: 0.875rem;
  }
}

.readOnlyTag {
  align-items: center;
  background: var(--cds-ui-01);
  border-radius: 0.875rem;
  display: flex;
  font-size: 0.875rem;
  height: 1.875rem;
  justify-content: center;
  letter-spacing: 0;
  margin-bottom: 0.75rem;
  width: 6.0625rem;
}

.readOnlyText {
  color: var(--cds-text-01);
  font-size: 0.875rem;
  letter-spacing: 0;
  line-height: 1.125rem;
  max-width: 42rem;
}

.settingsDescription {
  font-size: 0.875rem;
  line-height: 1.125rem;
  max-width: 41.25rem;
}


.sectionContainer {
  border-top: 0.0625rem solid var(--cds-active-ui);
  margin-top: 1.5rem;
  padding-top: 1.5rem;
  width: 100%;
}

.sectionHeader {
  align-items: center;
  display: flex;
}

.sectionTitle {
  font-size: 1.125rem;
  letter-spacing: 0;
  line-height: 1.4375rem;
  margin: 0 0.5rem 0 0;
}

.sectionDescription {
  font-size: 0.875rem;
  letter-spacing: 0.01rem;
  line-height: 1.125rem;
  margin-top: 1.25rem;
  width: 41.1875rem;
}


.buttonWithMessageContainer {
  margin-top: 1.5rem;
  width: 41.1875rem;
}

.buttonHelperText {
  color: var(--cds-text-02);
  font-size: 0.75rem;
  font-style: normal !important;
  line-height: 0.9375rem;
  margin-bottom: 1rem !important;
  max-width: 41.2rem;
}

.buttonClose {
  margin-left: -0.5rem;
  padding: 0.1875rem 0.5rem !important;
}