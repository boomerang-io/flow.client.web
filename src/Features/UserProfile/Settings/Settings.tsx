import React from "react";
import { Helmet } from "react-helmet";
import { ConfirmModal, notify, ToastNotification } from "@boomerang-io/carbon-addons-boomerang-react";
import { useMutation } from "react-query";
import { Edit, Close } from "@carbon/react/icons";
import { Tag, Button } from "@carbon/react";
import { FlowUser } from "Types";
import { resolver, serviceUrl } from "Config/servicesConfig";
import styles from "./Settings.module.scss";

interface UserSettingsProps {
  user: FlowUser;
  userManagementEnabled: any;
}

export default function Settings({ user, userManagementEnabled }: UserSettingsProps) {
  const canEdit = userManagementEnabled;

  const removeUserMutator = useMutation(resolver.deleteUser);

  const removeTeam = async () => {
    try {
      await removeUserMutator.mutateAsync({ userId: user.id });
      notify(
        <ToastNotification title="Close Account" subtitle="Request to close your account successful" kind="success" />
      );
    } catch (error) {
      notify(
        <ToastNotification
          title="Close Account"
          subtitle={`Unable to close the account. ${error.message}. Please contact support.`}
          kind="error"
        />
      );
    }
  };

  return (
    <section aria-label="User Settings" className={styles.settingsContainer}>
      <Helmet>
        <title>{`Settings - ${user?.name}`}</title>
      </Helmet>
      {!canEdit ? (
        <section className={styles.readOnly}>
          <Tag className={styles.readOnlyTag}>Read-only</Tag>
          <p className={styles.readOnlyText}>
            Manage your profile - You don’t have permission to change any of these settings, but you can still see
            what’s going on behind the scenes.
          </p>
        </section>
      ) : (
        <p className={styles.settingsDescription}>
          Manage your profile, activate special features, or close your account.
        </p>
      )}
      <SettingSection title="Basic details">
        <dl className={styles.detailedListContainer}>
          <div className={styles.detailedListGrid}>
            <div className={styles.detailedListGridItem}>
              <dt className={styles.detailedListTitle}>Name</dt>
              <dd className={styles.detailedListDescription}>{user.name}</dd>
            </div>
            <div className={styles.detailedListGridItem}>
              <dt className={styles.detailedListTitle}>Email</dt>
              <dd className={styles.detailedListDescription}>{user.email}</dd>
            </div>
          </div>
        </dl>
      </SettingSection>
      <SettingSection title="API Tokens">
        <div className={styles.buttonWithMessageContainer}>
          <p className={styles.buttonHelperText}>Coming soon.</p>
        </div>
      </SettingSection>
      <SettingSection title="Features">
        <div className={styles.buttonWithMessageContainer}>
          <p className={styles.buttonHelperText}>There are no special features to be enabled at this time.</p>
        </div>
      </SettingSection>
      <SettingSection title="Close Account">
        <div className={styles.buttonWithMessageContainer}>
          <p className={styles.buttonHelperText}>
            Done with your work here? Closing your account means you will no longer be able to access any Teams or
            Workflows you have created. You will also no longer receive any notifications from the platform.
          </p>
          <p className={styles.buttonHelperText}>
            This action cannot be undone. Be sure you want to permanently delete all of the data stored.
          </p>
          <ConfirmModal
            affirmativeAction={() => removeTeam()}
            affirmativeButtonProps={{ kind: "danger", "data-testid": "confirm-close-account" }}
            title="Close Account?"
            negativeText="Cancel"
            affirmativeText="Close"
            modalTrigger={({ openModal }: { openModal: () => void }) => (
              <Button
                disabled={!canEdit}
                iconDescription="Close"
                kind="danger"
                onClick={openModal}
                renderIcon={Close}
                size="md"
                data-testid="close-team"
              >
                Close Account
              </Button>
            )}
          >
            Closing your account will submit a request but will not immediatly become inactive. This action cannot be
            undone. Are you sure you want to do this?
          </ConfirmModal>
        </div>
      </SettingSection>
    </section>
  );
}

// function EditButton({ openModal }) {
//   return (
//     <Button
//       className={styles.editButton}
//       data-testid="settings-edit-button"
//       iconDescription="edit"
//       kind="ghost"
//       size="small"
//       renderIcon={Edit}
//       onClick={openModal}
//     />
//   );
// }

function SettingSection({ children, description, editModal, title }) {
  return (
    <section className={styles.sectionContainer}>
      <div className={styles.sectionHeader}>
        <h1 className={styles.sectionTitle}>{title}</h1>
        {editModal}
      </div>
      {description ? <p className={styles.sectionDescription}>{description}</p> : null}
      {children}
    </section>
  );
}
