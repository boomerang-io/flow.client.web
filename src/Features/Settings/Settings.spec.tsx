/* eslint-disable */
import { vi } from "vitest";
import Settings from "./Settings";
import { fireEvent, screen, waitFor } from "@testing-library/react";
import { Response } from "miragejs";
import { startApiServer } from "ApiServer";
import { serviceUrl } from "Config/servicesConfig";

let server: any;

beforeEach(() => {
  server = startApiServer();
});

afterEach(() => {
  server.shutdown();
});

describe("Settings --- Snapshot", () => {
  test("Capturing Snapshot of Settings", () => {
    const { baseElement } = global.rtlContextRouterRender(<Settings />);
    expect(baseElement).toMatchSnapshot();
  });
});

describe("Settings --- RTL", () => {
  test("Loads and opens section", async () => {
    global.rtlContextRouterRender(<Settings />);
    expect(screen.getByRole("heading", { name: /^Settings$/i })).toBeInTheDocument();
    const section = await screen.findByRole("heading", { name: "Workers" });
    fireEvent.click(section);
    expect(screen.getByLabelText(/^Enable Debug$/i)).toBeInTheDocument();
  });
});
describe("Settings --- RTL", () => {
  beforeEach(() => {
    server.get(serviceUrl.resourceSettings(), () => {
      return new Response(500, {}, {});
    });
  });
  test("Shows error message on request failure", async () => {
    global.rtlContextRouterRender(<Settings />);
    expect(screen.getByRole("heading", { name: /^Settings$/i })).toBeInTheDocument();
    await waitFor(() => {
      expect(screen.getByText("Oops, something went wrong.")).toBeInTheDocument();
    });
  });
});
