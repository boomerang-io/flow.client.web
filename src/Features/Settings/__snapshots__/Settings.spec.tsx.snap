// Vitest Snapshot v1

exports[`Settings --- Snapshot > Capturing Snapshot of Settings 1`] = `
<body
  id="app"
>
  <div>
    <header
      class="cds--bmrg-feature-header _header_10f51f"
    >
      <section
        class="cds--bmrg-feature-header__content"
      >
        <div
          class="cds--bmrg-feature-header__header"
        >
          <h1
            class="cds--bmrg-feature-header-title _headerTitle_10f51f"
          >
            Settings
          </h1>
          <p
            class="cds--bmrg-feature-header-title"
          >
            Adjust Flow settings
          </p>
        </div>
      </section>
    </header>
    <div
      class="_container_10f51f css-xtq108"
    >
      <div
        class="_skeletonPlacholderContainer_10f51f css-vurnku"
      >
        <div
          class="cds--skeleton__placeholder"
        />
        <div
          class="cds--skeleton__placeholder"
        />
        <div
          class="cds--skeleton__placeholder"
        />
        <div
          class="cds--skeleton__placeholder"
        />
        <div
          class="cds--skeleton__placeholder"
        />
      </div>
    </div>
  </div>
</body>
`;
